### PR TITLE
drivers: sensor: icm42670: supports icm42670-P and icm42670-S

### PR DESCRIPTION
**Pull request by afontaine-invn**
dts: supports icm42670-P/-S
samples: sensor: tdk_apex: TDK APEX generic sample
samples: sensor: 6DOF motion DRDY generic sample
module: HAL TDK module

Adds official TDK Invensense Inc. driver in TDK HAL  module for icm42670-P/-S sensor.
Adds SPI and I2C interface support.
Adds APEX features, such as Pedometer, Tilt detection, Wake on Motion
and Significant Motion Detector.
Add generic samples supporting this driver.
Integration module test successful with https://github.com/tdk-invn-oss/zephyr.hal_tdk  
Validated with custom setup: nrf52dk_nrf52832 + icm42670-P/-S EVB board
Build ok by running:
west twister -T samples\sensor\tdk_apex-T samples\sensor\6dof_motion_drdy -p nrf52dk/nrf52832

Signed-off-by: Aurelie Fontaine <aurelie.fontaine@tdk.com>

### CODE REVIEW

**Review by gmarull — CHANGES_REQUESTED** [review]
please, check Zephyr contribution guidelines. There seem to be PDFs and binary files in this PR.

**Review by teburd — COMMENTED** [review]
An initial pass here (and this is large so I'll need time to look at it again) with some high level notable comments but...

I love that this is being submitted as a PR from you, with a TDK email address. I hope this is just the first of many to be frank.

That said there's some issues that will need to be addressed. Blobs are problematic, pdf's are problematic, rewriting the register map but not using the zephyr macros for describing masks and manipulating fields seems like a step backwards. Using the existing trigger API requiring a thread or work queue also feels like a step backwards, but that's not on you.

Is the code under the imu directory here something from a TDK software devkit? If so it might be worth considering looking at how ST packages and uses stmemsc for ST sensor drivers. It does not need to live or be managed then by Zephyr but can be managed entirely but TDK in something like a hal_tdk git repository.

**Comment by carlescufi** [comment]
Thanks for the PR! Please take a look at [our commit message guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines) to find out how to format your commit messages, and at [our contribution workflow](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow) to understand how to update your Pull Request.

**Comment by afontaine-invn** [comment]
Hi @teburd, @carlescufi, thank you very much for your review! You point many subjects to address, Find my answers below: 

- Regarding PDF and blobs I'll remove it for now. We though exposing the AML library (featuring pointing, gesture recognition, ...) but it seems not feasible and compatible with Zephyr rules.
- Regarding the trigger API requiring a thread or work queue I kept it configurable by KConfig and I kept the option to not configure any trigger. I can put it as default.

- Regarding the possibility to use a hal_tdk git repository to put the imu driver directory, we look at how st handles it with @gjabouley-invn and we'll like the idea. Could you give us more details on how to put it in place? And if there is any rules or guidelines to manage it over the time?

- And just as you know I'm not the only one to push with a TDK address lately, @
rbuisson-invn also updated ICP drivers with official ones. Find the PR https://github.com/zephyrproject-rtos/zephyr/pull/69282

File: dts/bindings/sensor/invensense,icm42670-i2c.yaml
Comment (decsny) [inline dts/bindings/sensor/invensense,icm42670-i2c.yaml:1]: any DT binding property with a default value needs justification of the default value in the description https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html

**Review by decsny — DISMISSED** [review]

File: drivers/sensor/tdk/icm42670/invn.algo.sw-integration.aml-42670s-gcc-arm-none-eabi-cm4-fpu-1.6.0/include/invn_algo_aml.h
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/invn.algo.sw-integration.aml-42670s-gcc-arm-none-eabi-cm4-fpu-1.6.0/include/invn_algo_aml.h]: This file has been removed

File: samples/sensor/icm42670/data_stream/src/main.c
Comment (afontaine-invn) [reply samples/sensor/icm42670/data_stream/src/main.c]: Sample moved to 6dof_motion_drdy as you suggested.

File: samples/sensor/icm42670/apex/src/main.c
Comment (afontaine-invn) [reply samples/sensor/icm42670/apex/src/main.c]: Sample moved to tdk_apex as you suggested.

File: drivers/sensor/tdk/icm42670/invn.algo.sw-integration.aml-42670s-gcc-arm-none-eabi-cm4-fpu-1.6.0/include/invn_algo_aml.h
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/invn.algo.sw-integration.aml-42670s-gcc-arm-none-eabi-cm4-fpu-1.6.0/include/invn_algo_aml.h]: This file has been removed

File: dts/bindings/sensor/invensense,icm42670-i2c.yaml
Comment (afontaine-invn) [reply dts/bindings/sensor/invensense,icm42670-i2c.yaml:1]: I updated the DT bindings with default description as required..

File: dts/bindings/sensor/invensense,icm42670-common.yaml
Comment (decsny) [inline dts/bindings/sensor/invensense,icm42670-common.yaml:17]: missing justification

Comment (afontaine-invn) [reply dts/bindings/sensor/invensense,icm42670-common.yaml:17]: Thanks for the comment. I added the default value justification for this property also.

**Comment by teburd** [comment]
Sensor WG 04/22/2024

Asked the question around pros/cons of using sensor vendor supplied HALs:

Pros:
* Potentially more vendor involvement, with more drivers being pushed to the tree

Potential Cons:
* Unclear how much benefit the HAL provides, e.g. abstracting bus reads/writes is already doable within Zephyr today.
* Unclear how the maintainer ship burden increases when using HALs, who maintains this code long term and moves it forward as sensor APIs evolve and change.
* Unclear how we move this code to the asynchronous sensor_read API which is the future of sensors to better support more use cases

**Comment by afontaine-invn** [comment]
Hi @teburd ,
Thanks for the updates on the tdk_hal repository issue.
Are there any blocking points or any changes expected to move forward with the PR?

**Comment by teburd** [comment]
> Hi @teburd , Thanks for the updates on the tdk_hal repository issue. Are there any blocking points or any changes expected to move forward with the PR?

No blocking issues from me no, I expect the TSC will approve hal_tdk barring any major concerns, and then this can go in (as it depends on hal_tdk). I'd expect some more progress in a week or so as the TSC meets weekly, and our sensor maintainer (Maureen Helm at Analog) is traveling this week

File: drivers/sensor/tdk/icm42670/Invn/InvError.h
Comment (gmarull) [inline drivers/sensor/tdk/icm42670/Invn/InvError.h:12]: this file still doesn't follow Zephyr conventions (doxygen not used in internal headers, UpperCase naming, etc.)

File: drivers/sensor/tdk/icm42670/Invn/InvError.h
Comment (gmarull) [inline drivers/sensor/tdk/icm42670/Invn/InvError.h:29]: not sure what these are? why not use `errno` codes?

Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/Invn/InvError.h:26]: Thank you for the review. 
This InvError.h file is part of our driver package (that should move to be in hal_tdk module). I removed any use of them in icm42670.c file.

File: drivers/sensor/tdk/icm42670/Invn/InvError.h
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/Invn/InvError.h]: Here I removed the @ingroup comments not relevant in this integration. Are there other changes you requested?

**Comment by ubieda** [comment]
@afontaine-invn would you like to join the Sensors WG meeting on monday to touch base on this PR?
It's next monday 7 to 8 AM PT. Here are the details: https://github.com/zephyrproject-rtos/zephyr/wiki/Sensors-Working-Group

**Comment by ubieda** [comment]
> Hi @ubieda , Hi @teburd , Following last TSC meeting, I updated the PR to move all TDK drivers in TDK Hal module. I tested the module integration successfully with [tdk-invn-oss/zephyr.hal_tdk](https://github.com/tdk-invn-oss/zephyr.hal_tdk) copied in module/hal/tdk I updated the west.yml as if the TDK HAL module is already a zephyr repository. Let me know if it's ok as it breaks the build. Or if there is another process?

Thanks for following up!

The process for submitting external code (hal_tdk) as a Module is described here: https://docs.zephyrproject.org/latest/contribute/external.html#submission-and-review-process.

From what I see, your next steps would include:
- Create the GH Issue following the template. Within the information filled, you would link this PR as part of the submission (see this reference: #72391).
- Present this module in the next TSC meeting, where voting takes place to decide on its inclusion after review. These meetings take place on Wednesdays from 8 to 9 AM PT (See: https://github.com/zephyrproject-rtos/zephyr/wiki/Technical-Steering-Committee-%28TSC%29).
- Once the module is approved, the Zephyr Project creates the `zephyrproject-rtos/hal_tdk` repository, managed by the Zephyr Project, maintained and kept up to date by the TDK team.
- That unlocks this PR review to finalize, assuming the contents pertintent to Zephyr are here and the external code is offloaded to the HAL (at a glance it already appears to be the case).

Feel free to reach out if you have any further questions.

@teburd @nashif Please, LMK if I missed anything.

**Comment by afontaine-invn** [comment]
Thanks for the process reminder!
The issue is opened: https://github.com/zephyrproject-rtos/zephyr/issues/71561
I'll join the meeting as next step!

**Comment by LucasTambor** [comment]
@afontaine-invn I just pushed #74418 before I found out about this PR. I tried to be as minimally invasive as possible to enable the I2C bus for the ICM42670. 

This PR addresses all my needs in addition to adding other features. Do you think is better to close my PR, or we can proceed on approving it before yours, since it is smaller and is likely to get in before this one?

**Comment by ubieda** [comment]
> Do you think is better to close my PR, or we can proceed on approving it before yours, since it is smaller and is likely to get in before this one?

I'd vote for pursuing this one instead.

**Comment by ubieda** [comment]
@LucasTambor Added you to the reviewer's list per https://github.com/zephyrproject-rtos/zephyr/pull/74418#issuecomment-2176687238

**Comment by afontaine-invn** [comment]
> Thanks for the process reminder! The issue is opened: #71561 I'll join the meeting as next step!

HI @ubieda , Do you know what I need to do now that the motion is passed on the last 10th July TSC meeting?

**Comment by afontaine-invn** [comment]
Hi @nashif , as I understand the tests doesn't pass because of a missing test in CMakelist in hal_tdk, meaning part of the files are included.  https://github.com/zephyrproject-rtos/hal_tdk/pull/1. How can I push this change?

**Comment by afontaine-invn** [comment]
Hi @ubieda , as I understand the tdk_hal module I added breaks the test. I would like to fix it https://github.com/zephyrproject-rtos/hal_tdk/pull/1. Could you help reviewing it to unlock the build in this PR?

**Comment by afontaine-invn** [comment]
> please, check Zephyr contribution guidelines. There seem to be PDFs and binary files in this PR.

Hi @gmarull , the PDF and binary files has been removed, could you have a look at this PR? Thanks

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (gmarull) [inline dts/bindings/sensor/invensense,icm42670.yaml:68]: no spaces in enums please

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (gmarull) [inline dts/bindings/sensor/invensense,icm42670.yaml:102]: units?

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (gmarull) [inline dts/bindings/sensor/invensense,icm42670.yaml:134]: units?

File: samples/sensor/6dof_motion_drdy/src/main.c
Comment (gmarull) [inline samples/sensor/6dof_motion_drdy/src/main.c:30]: this is impossible, build will fail if macro doesn't resolve

File: samples/sensor/tdk_apex/VERSION
Comment (gmarull) [inline samples/sensor/tdk_apex/VERSION:5]: ??

File: samples/sensor/tdk_apex/src/main.c
Comment (gmarull) [inline samples/sensor/tdk_apex/src/main.c:31]: same

**Review by gmarull — CHANGES_REQUESTED** [review]

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (afontaine-invn) [reply dts/bindings/sensor/invensense,icm42670.yaml]: I replaced space by -

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (afontaine-invn) [reply dts/bindings/sensor/invensense,icm42670.yaml:102]: Unit is Hz, I added it in the comment.

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (afontaine-invn) [reply dts/bindings/sensor/invensense,icm42670.yaml:134]: Unit is Hz, I added it in the comment.

File: samples/sensor/6dof_motion_drdy/src/main.c
Comment (afontaine-invn) [reply samples/sensor/6dof_motion_drdy/src/main.c]: Ok I removed it. I copied this from the bosh,bme280 example, I though it was needed.

File: samples/sensor/tdk_apex/src/main.c
Comment (afontaine-invn) [reply samples/sensor/tdk_apex/src/main.c]: Ok I removed it.

File: samples/sensor/tdk_apex/VERSION
Comment (afontaine-invn) [reply samples/sensor/tdk_apex/VERSION]: I removed the VERSION files. I though these VERSION files were recommended, but I misunderstood the guidelines.

**Comment by afontaine-invn** [comment]
Hi @gmarull , I did the changes you requested. Could you update the changes requested status if it's ok for you?

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (gmarull) [reply dts/bindings/sensor/invensense,icm42670.yaml:102]: property still misses proper suffix

Comment (afontaine-invn) [reply dts/bindings/sensor/invensense,icm42670.yaml:102]: I don't get your point. Which suffix is missing for accel-filt-bw?

**Comment by afontaine-invn** [comment]
Hello @gmarull 
I addressed the change as requested; could you update your review status? Is there something else do you expect?

File: samples/sensor/6dof_motion_drdy/README.rst
Comment (gmarull) [inline samples/sensor/6dof_motion_drdy/README.rst:4]: wrong formatting

File: samples/sensor/6dof_motion_drdy/boards/nrf52dk_nrf52832_i2c.overlay
Comment (gmarull) [inline samples/sensor/6dof_motion_drdy/boards/nrf52dk_nrf52832_i2c.overlay:42]: why is this needed?

**Review by gmarull — DISMISSED** [review]

File: samples/sensor/6dof_motion_drdy/README.rst
Comment (afontaine-invn) [reply samples/sensor/6dof_motion_drdy/README.rst]: I updated both README for 6dof_motion_drdy and tdk_apex samples.

File: samples/sensor/tdk_apex/Kconfig
Comment (afontaine-invn) [reply samples/sensor/tdk_apex/Kconfig]: Ok, I moved this to the prj.conf file

File: drivers/sensor/tdk/icm42670/Kconfig
Comment (pdgendt) [inline drivers/sensor/tdk/icm42670/Kconfig:33]: Omit `default n`, but it feels like this should be a dts/runtime setting. What if you have multiple sensors, and want each sensor to have a different feature?

File: drivers/sensor/tdk/icm42670/Kconfig
Comment (pdgendt) [inline drivers/sensor/tdk/icm42670/Kconfig:29]: This should go to device tree instead of Kconfig, imagine having multiple variants.

File: drivers/sensor/tdk/icm42670/Kconfig
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/Kconfig]: You are right, it's better to put it in device tree. I did the change and the result is cleaner, thanks!

File: drivers/sensor/tdk/icm42670/Kconfig
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/Kconfig]: Thanks for this comment as well. I did the change and put this option in device tree. It has more sense like this.

**Comment by afontaine-invn** [comment]
Thanks for your comment @pdgendt , it was pertinent and I'm happy with the new version with more options in device tree.

**Comment by afontaine-invn** [comment]
Hi @gmarull , I think the changes I made match what you expected. Could you review and remove the changes requested that may block the PR ?

**Comment by afontaine-invn** [comment]
Hi @MaureenHelm , could we remove the DNM label on this PR ? I did a lot of changes to match what was expected from the Zephyr members comments and I would like to move forward ?

File: drivers/sensor/tdk/icm42670/icm42670.c
Comment (MaureenHelm) [inline drivers/sensor/tdk/icm42670/icm42670.c:729]: Look at implementing the sensor streaming interface with RTIO

File: drivers/sensor/tdk/icm42670/icm42670.c
Comment (MaureenHelm) [inline drivers/sensor/tdk/icm42670/icm42670.c:814]: This is another example where you will end up returning something that isn't an errno anymore because you've logically OR-ed together multiple values

File: drivers/sensor/tdk/icm42670/icm42670_trigger.h
Comment (MaureenHelm) [inline drivers/sensor/tdk/icm42670/icm42670_trigger.h:17]: ???

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (MaureenHelm) [inline dts/bindings/sensor/invensense,icm42670.yaml:22]: Use different compatibles instead

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (MaureenHelm) [inline dts/bindings/sensor/invensense,icm42670.yaml:1]:
```
dts: variant, apex, power, and filtering options

Add the power mode, accel and gyro filtering options, variant S
support, and apex features.

Signed-off-by: Aurelie Fontaine <aurelie.fontaine@tdk.com>
```

The first line of the commit message is too vague, can you update it to something like:
```dts: bindings: sensor: Add invensense icm42670 properties```

This makes it easier to summarize sensor-related changes for release notes

File: samples/sensor/6dof_motion_drdy/Kconfig
Comment (MaureenHelm) [inline samples/sensor/6dof_motion_drdy/Kconfig:7]: This file is for defining new Kconfig symbols, not overriding existing symbols that were defined elsewhere. You should set this in prj.conf instead.

File: samples/sensor/6dof_motion_drdy/README.rst
Comment (MaureenHelm) [inline samples/sensor/6dof_motion_drdy/README.rst:1]: @avisconti can you take a look at this sample?

File: samples/sensor/6dof_motion_drdy/boards/nrf52dk_nrf52832_i2c.overlay
Comment (MaureenHelm) [inline samples/sensor/6dof_motion_drdy/boards/nrf52dk_nrf52832_i2c.overlay:17]: Is there an arduino shield available with this sensor?

File: samples/sensor/tdk_apex/Kconfig
Comment (MaureenHelm) [inline samples/sensor/tdk_apex/Kconfig:7]: Ditto

File: samples/sensor/tdk_apex/README.rst
Comment (MaureenHelm) [inline samples/sensor/tdk_apex/README.rst:64]:
```suggestion
.. zephyr-app-commands::
```

File: west.yml
Comment (MaureenHelm) [inline west.yml:244]: The HAL commit should be first in the series, otherwise this PR isn't bisectable

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: samples/sensor/6dof_motion_drdy/boards/nrf52dk_nrf52832_i2c.overlay
Comment (afontaine-invn) [reply samples/sensor/6dof_motion_drdy/boards/nrf52dk_nrf52832_i2c.overlay:17]: No, there is no Arduino shield available for now.

File: drivers/sensor/tdk/icm42670/Kconfig
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/Kconfig]: Change done

File: drivers/sensor/tdk/icm42670/icm42670.c
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/icm42670.c]: Thanks for the catch. I updated the error management to report only errno's type to the sensor driver API.

File: drivers/sensor/tdk/icm42670/icm42670.c
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/icm42670.c]: Change done

File: drivers/sensor/tdk/icm42670/icm42670.c
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/icm42670.c]: Thanks for this catch. It's now handled according to the instance in Kconfig.

File: drivers/sensor/tdk/icm42670/icm42670.c
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/icm42670.c]: Change done

File: drivers/sensor/tdk/icm42670/icm42670.c
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/icm42670.c]: Change done

File: drivers/sensor/tdk/icm42670/icm42670_trigger.h
Comment (afontaine-invn) [reply drivers/sensor/tdk/icm42670/icm42670_trigger.h]: Sorry, formatting error on my side. Change reverted.

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (afontaine-invn) [reply dts/bindings/sensor/invensense,icm42670.yaml]: Thanks for this suggestion, I created different compatible.

File: dts/bindings/sensor/invensense,icm42670.yaml
Comment (afontaine-invn) [reply dts/bindings/sensor/invensense,icm42670.yaml:1]: Commit message updated

File: samples/sensor/6dof_motion_drdy/Kconfig
Comment (afontaine-invn) [reply samples/sensor/6dof_motion_drdy/Kconfig]: Thanks for this clarification. I removed the Kconfig and put this in prj.conf

File: samples/sensor/tdk_apex/Kconfig
Comment (afontaine-invn) [reply samples/sensor/tdk_apex/Kconfig]: Move to prj.conf as requested

File: samples/sensor/tdk_apex/README.rst
Comment (afontaine-invn) [reply samples/sensor/tdk_apex/README.rst]: Change done

File: west.yml
Comment (afontaine-invn) [reply west.yml:244]: Thanks for this comment. I reorder my commits to have the HAL commit first.

**Comment by sriccardi-invn** [comment]
HI @MaureenHelm,
Any update on the MR ?
The branch is opened since a lot of time with different reviewers. My team consume is time to rebase everytime.
Do you think we could converge?
We have already a lot of sensor drivers ready to be integrated in tdk_hal.
Thanks.
Regards,
Sebastien Riccardi

**Review by MaureenHelm — DISMISSED** [review]

**Comment by afontaine-invn** [comment]
Hello @MaureenHelm , Thanks for the approval, we're happy to achieve this step!
I rebased this PR and resolved the conflict in icm42670.c file. Could you approve again and remove the DNM and dev-review label in order to be able to merge?

**Comment by sriccardi-invn** [comment]
HI @MaureenHelm ,
I am affraid by the delay to merge a branch.
Everytime we wait, some change have beed done an another side, we have ro rebase, wait a new validation.
This is infinite.
We have a lot of sensors driver ready to be integrated waiting this first merge.
Do you have or see any remaining issue on your side?
Thanks.
Regards,
Sebastien Riccardi

**Review by teburd — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Comment by kartben** [comment]
Thank you for your patience @afontaine-invn @sriccardi-invn and team, and sorry this took so long. This will go in with the next merge batch in just a couple hours.

**Review by kartben — APPROVED** [review]

**Comment by pdgendt** [comment]
Every build now prints:

```
CMake Warning at /home/pdgendt/dev/zephyr-main/zephyr/CMakeLists.txt:996 (message):
  No SOURCES given to Zephyr library: ..__modules__hal__tdk

  Excluding target from build.
```

**Comment by pdgendt** [comment]
Created https://github.com/zephyrproject-rtos/zephyr/pull/83151 to get rid of the warning.

**Pull request merged**
